File: services/typescript/fastify/src/routes/hooks/chaos.ts
import type { onRequestAsyncHookHandler } from "fastify";
import { randomInt } from "node:crypto";
import { setTimeout } from "node:timers/promises";
import type { State } from "app";
import { AppErrors, sendError } from "../errors.js";

export function chaosHook(state: State): onRequestAsyncHookHandler {
  return async (_request, reply) => {
    const config = state.config;

    if (config.latency.enabled && randomInt(0, 101) < config.latency.rate) {
      // Note: can't do sub ms but logic is same
      await setTimeout(randomInt(500, 1501) / 1000);
    }
    if (config.error.enabled && randomInt(0, 101) < config.error.rate) {
      return sendError(reply, AppErrors.Internal);
    }
  };
}

File: services/python/sanic/src/routes/hooks/chaos.py
from asyncio import sleep
from random import randrange
from sanic import Request
from sanic.response import JSONResponse

from routes.errors import send_error, AppErrors


async def chaos_hook(
    request: Request,
) -> JSONResponse | None:
    config = request.app.ctx.config

    if config.latency.enabled and randrange(0, 101) < config.latency.rate:
        await sleep(randrange(500, 1501) / 1_000_000)
    if config.error.enabled and randrange(0, 101) < config.error.rate:
        return send_error(AppErrors.Internal)
